pkg/graveler/ref/manager.go

```diff
@@ -412,19 +445,78 @@ func (m *Manager) ListCommits(ctx context.Context, repositoryID graveler.Reposit
-									FROM graveler_commits c INNER JOIN cte
-									ON cte.id = ANY(c.parents) WHERE repository_id = $1
-								)
-								UPDATE graveler_commits u SET generation = cte.generation FROM cte WHERE u.id = cte.id;`,
-			repositoryID)
+	// update commitNodes' generation in nodes "tree" using BFS algorithm.
+	// using a queue implementation
+	// adding a node to the queue only after all of its parents were visited in order to avoid redundant visits of nodesCommitsIDs
+	nodes, err := m.createCommitsIDsMap(ctx, repositoryID)
+	if err != nil {
+		return err
+	}
+	rootsCommitIDs := m.getRootNodes(nodes)
+	m.mapCommitNodesToChildren(nodes)
+	m.addGenerationToNodes(nodes, rootsCommitIDs)
+	_, err = m.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
+		return nil, m.updateCommitGeneration(tx, repositoryID, nodes)
 	})
 	return err
 }
+
+func (m *Manager) createCommitsIDsMap(ctx context.Context, repositoryID graveler.RepositoryID) (map[graveler.CommitID]*CommitNode, error) {
+	iter, err := m.ListCommits(ctx, repositoryID)
+	if err != nil {
+		return nil, err
+	}
+	defer iter.Close()
+
+	nodes := make(map[graveler.CommitID]*CommitNode)
+	for iter.Next() {
+		commit := iter.Value()
+		parentsToVisit := map[graveler.CommitID]struct{}{}
+		for _, parentID := range commit.Parents {
+			parentsToVisit[parentID] = struct{}{}
+		}
+
+		nodes[commit.CommitID] = &CommitNode{parentsToVisit: parentsToVisit}
+	}
+	if err := iter.Err(); err != nil {
+		return nil, fmt.Errorf("getting value from iterator: %w", err)
+	}
+	return nodes, nil
+}
+
+func (m *Manager) getRootNodes(nodes map[graveler.CommitID]*CommitNode) []graveler.CommitID {
+	var rootsCommitIDs []graveler.CommitID
+	for commitID, commit := range nodes {
+		if len(commit.parentsToVisit) == 0 {
+			rootsCommitIDs = append(rootsCommitIDs, commitID)
+		}
+	}
+	return rootsCommitIDs
+}
+
+func (m *Manager) mapCommitNodesToChildren(nodes map[graveler.CommitID]*CommitNode) {
+	for commitID, commitNode := range nodes {
+		// adding current node as a child to all parents in commitNode.parentsToVisit
+		for parentID := range commitNode.parentsToVisit {
+			nodes[parentID].children = append(nodes[parentID].children, commitID)
+		}
+	}
+}
+
+func (m *Manager) addGenerationToNodes(nodes map[graveler.CommitID]*CommitNode, rootCommits []graveler.CommitID) {
+	nodesCommitsIDs := rootCommits
+	for currentGeneration := 1; len(nodesCommitsIDs) > 0; currentGeneration++ {
+		var nextIterationNodes []graveler.CommitID
+		for _, nodeCommitID := range nodesCommitsIDs {
+			currentNode := nodes[nodeCommitID]
+			nodes[nodeCommitID].generation = currentGeneration
+			for _, childNodeID := range currentNode.children {
+				delete(nodes[childNodeID].parentsToVisit, nodeCommitID)
+				if len(nodes[childNodeID].parentsToVisit) == 0 {
+					nextIterationNodes = append(nextIterationNodes, childNodeID)
+				}
+			}
+		}
+		nodesCommitsIDs = nextIterationNodes
+	}
+}
```

Comment:
> Thanks!
> 
> Looks very nice. One SQL injection vector remains and blocks approval.
> 
> I would also be _happy_ if you got rid of the map in favour of a reference count. Right now I do not think it should block us; however if people do a commit every 10s then it may become significant. In any case we can do it on a separate PR.
> 
> Similarly, I'd really like for you to test the generations computation. That too does not need to happen in this PR.

@arielshaqed about the change of `parentsToVisit` to be counter, I'm using this map in `mapCommitNodesToChildren` so I can't really change it to be a counter